pkg/scheduler/eventhandlers.go

```diff
@@ -243,6 +243,15 @@ func (sched *Scheduler) updatePodInCache(oldObj, newObj interface{}) {
 		return
 	}
 
+	// A Pod delete event followed by an immediate Pod add event may be merged
+	// into a Pod update event. In this case, we should invalidate the old Pod, and
+	// then add the new Pod.
+	if oldPod.UID != newPod.UID {
```

Comment:
Why are we here? shouldn't we be here only if NodeName was set? it is not going to be set for the new pod, right?

Reply:
It's described in #85726 in a HA setup and the author provided a detailed log. Basically it's like this:

- schedA as the master, schedB as the standby.
- testPod is scheduled to node1; schedA and schedB both gets that cached properly.
- shut down schedB.
- at the same time, testPod is deleted and then created - schedA works fine - 2 events (Delete and Add) get received, and the new testPod got scheduled (to node1 let's say).
- shut down schedA and start schedB.
- an interesting thing happens. schedB receives 1 Update event (instead of 2 events) on testPod. Because schedB only has the stale testPod's UID in place, and the internal scheduler `cache state` uses UID to check the Pod's presence, it throws an error and failed the update:
    ```
    E1128 16:42:11.731020 31593 eventhandlers.go:226] scheduler cache UpdatePod failed: pod 3c6696c6-decb-4abd-a1d3-d891e032459d is not added to scheduler cache, so cannot be updated
    ```
- this error is critical, as it means the stale testPod live in scheduler's cache (instead of the new testPod) is corrupted.
- deleting testPod doesn't help, as it cannot be deleted due to UID not found.
    ```
    E1128 16:42:20.632578 31593 eventhandlers.go:254] scheduler cache RemovePod failed: pod 3c6696c6-decb-4abd-a1d3-d891e032459d is not found in scheduler cache, so cannot be removed from it
    ```

Reply:
ah, ok, so testPod did get scheduled a second time already, and so it's NodeName is set and that triggers a proper update.

I was assuming that the new pod isn't scheduled yet, and so according to the [OnUpdate handler](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/controller.go#L264-L276) the update should change to a delete event on the cache and an add event on the queue, which make sense.

Reply:
Is there a bug filled for kube-apiserver? I'm fine with dealing with it here for now.

Reply:
> Is there a bug filled for kube-apiserver?

I _prefer_ to believe it's a bug, but seems it's not the case from api's perspective, and I don't it can be "fixed" b/c the underlying DeltaFIFO queue is using namespace/name as the index.